### PR TITLE
podman-remote.conf enablement

### PR DESCRIPTION
**Pull request by baude**
add the ability for the podman remote client to use a configuration file
which describes its connections. users can now define a connection the
configuration and then call it by name like:

podman-remote -c connection1

and the destination and user will be derived from the configuration
file.  if no -c is provided, we look for a connection in the
configuration file designated as 'default'.  If the configuration file
has only one connection, it will be deemed the 'default'.

Signed-off-by: baude <bbaude@redhat.com>

### CODE REVIEW

File: cmd/podman/remoteclientconfig/config_linux.go [hunk 1]
```diff
@@ -0,0 +1,8 @@
+package remoteclientconfig
+
+import (
+	"os"
+	"path/filepath"
+)
+
+var configFilePath = filepath.Join(os.Getenv("HOME"), ".config", "containers", remoteConfigFileName)
```
Comment (jwhonce) [inline cmd/podman/remoteclientconfig/config_linux.go]: HOME is not portable, even across "unix'es"

`os.LookupEnv(XDG_CONFIG_HOME)` should be first check.
If needed, then
```
usr, _ := user.Current()
dir := usr.HomeDir
```

File: docs/podman-remote.conf.5.md [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+
+## DESCRIPTION
+The libpod.conf file is the default configuration file for all tools using
+libpod to manage containers.
+
+The podman-remote.conf file is the default configuration file for the podman
+remote client.  It is in the TOML format.  It is primarily used to keep track
+of the user's remote connections.
+
+## CONNECTION OPTIONS
```
Comment (jwhonce) [inline docs/podman-remote.conf.5.md]: What if I want to give the whole bridge string?

File: pkg/adapter/client.go [hunk 3]
```diff
@@ -6,47 +6,57 @@ import (
...
 
-	// I'm leaving this here for now as a document of the birdge format.  It can be removed later once the bridge
-	// function is more flushed out.
-	// bridge := `ssh -T root@192.168.122.1 "/usr/bin/varlink -A '/usr/bin/podman varlink \$VARLINK_ADDRESS' bridge"`
-	if len(r.cmd.RemoteHost) > 0 {
-		// The user has provided a remote host endpoint
+	// If the user defines an env variable for podman_varlink_bridge
+	// we use that as passed.
+	if bridge := os.Getenv("PODMAN_VARLINK_BRIDGE"); bridge != "" {
+		logrus.Debug("creating a varlink bridge based on env variable")
```
Comment (jwhonce) [inline pkg/adapter/client.go]: Message could include connection string to aid in debugging.  Same below.

File: pkg/adapter/client_unix.go [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+
+	if len(formattedBridge) < 1 && remoteConn == nil {
+		return nil, errors.New("bridge connections must either be created by string or remoteconnection")
+	}
+	if len(formattedBridge) > 0 {
+		endpoint.Connection = formattedBridge
+		return &endpoint, nil
+	}
+	endpoint.Connection = fmt.Sprintf(
+		`ssh -T %s@%s /usr/bin/varlink -A \'/usr/bin/podman --log-level=%s varlink \\\$VARLINK_ADDRESS\' bridge`,
```
Comment (jwhonce) [inline pkg/adapter/client_unix.go]: `ssh -T %s@%s ` should be `ssh -T %s@%s -- `

File: pkg/adapter/client.go [hunk 5]
```diff
@@ -6,47 +6,57 @@ import (
...
+		//	last resort is to make a socket connection with the default varlink address for root user
+	} else {
+		logrus.Debug("creating a varlink address based default root address")
+		remoteEndpoint, err = newSocketConnection(DefaultAddress)
 	}
 	return
 }
 
-// Connect provides a varlink connection
+// Connect provides a varlinit bbaude@REDHAT.COMnk connection
```
Comment (jwhonce) [inline pkg/adapter/client.go]: Remove email address from string

File: docs/podman-remote.conf.5.md (hunk 2 above)
Comment (baude) [reply docs/podman-remote.conf.5.md]: i assume you mean via env variable? then that overrules config file usage.

File: pkg/adapter/client.go (hunk 5 above)
Comment (baude) [reply pkg/adapter/client.go]: lol, that is a "hot" key from my keyboard.

File: docs/podman-remote.conf.5.md (hunk 2 above)
Comment (jwhonce) [reply docs/podman-remote.conf.5.md]: You don't allow me to mix bridge strings and host ip's for different hosts in same config file.

File: pkg/adapter/client.go (hunk 5 above)
Comment (jwhonce) [reply pkg/adapter/client.go]: Kerberos strikes again

File: docs/podman-remote.conf.5.md (hunk 2 above)
Comment (baude) [reply docs/podman-remote.conf.5.md]: no

File: cmd/podman/remoteclientconfig/config_linux.go (hunk 1 above)
Comment (rhatdan) [reply cmd/podman/remoteclientconfig/config_linux.go]: Remember this has to work on Windows also.

File: cmd/podman/remoteclientconfig/configfile.go [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+	return &remoteConfig, err
+}
+
+// GetDefault returns the default RemoteConnection. If there is only one
+// connection, we assume it is the default as well
+func (r *RemoteConfig) GetDefault() (*RemoteConnection, error) {
+	if len(r.Connections) == 0 {
+		return nil, ErrNoDefinedConnections
+	}
+	for _, v := range r.Connections {
```
Comment (rhatdan) [inline cmd/podman/remoteclientconfig/configfile.go]: I would prefer
```
		if len(r.Connections) == 1 {
			// if there is only one defined connection, we assume it is
			// the default whether tagged as such or not
			return &r.Connections[0], nil
		}
          	for _, v := range r.Connections {
...
```

File: cmd/podman/remoteclientconfig/config_linux.go (hunk 1 above)
Comment (baude) [reply cmd/podman/remoteclientconfig/config_linux.go]: it seems os.UserHomeDir is preferred now -> https://golang.org/pkg/os/#UserHomeDir

it should also be noted that this is for unix only ... windows I believe will be different and I frankly don't know where it should go in that case.

Comment (rhatdan) [reply cmd/podman/remoteclientconfig/config_linux.go]: According to that link Windows would do the right thing as well.
```
On Unix, including macOS, it returns the $HOME environment variable. On Windows, it returns %USERPROFILE%. On Plan 9, it returns the $home environment variable. 
```

Comment (baude) [reply cmd/podman/remoteclientconfig/config_linux.go]: yeah but the more interesting question is where in window's dir structure should it go?

**Comment by rhatdan** [comment]
I don't want this config file to move into a general podman.conf file, Just for podman-remote connections.
I don't mind if it handles any flags that can be set in podman-remote.

**Comment by rhatdan** [comment]
LGTM

**Pull request merged**